mscore/plugin/pluginUpdater.cpp

```diff
@@ -0,0 +1,994 @@
+      if (should_update) {
+            desc.update = desc_tmp;
+            r->commitPlugin(page_url, desc);
+            emit pluginStatusChanged(install->property("row").toInt(), PluginStatus::UPDATE_AVAILABLE);
+            }
+      else {
+            delete desc_tmp;
+            emit pluginStatusChanged(install->property("row").toInt(), PluginStatus::UPDATED);
+            }
+      }
+
+void PluginWorker::timeconsume()
+      {
+      qDebug("timeconsume start");
+      int member = 0;
+      for (int i = 0; i < 1000000; i++)
+            member = member * 2 + 1;
+      qDebug("timeconsume end");
+      emit finished();
+      }
+
+bool PluginWorker::analyzePluginPage(QString full_url)
+      {
+      bool new_plugin = desc.source == UNKNOWN;
+      QString direct_link;
+      DownloadUtils page(QNetworkRequest::RedirectPolicy::NoLessSafeRedirectPolicy);
+      page.setTarget(full_url);
+      page.download();
+      QByteArray html_raw = page.returnData();
+      std::vector<PluginPackageLink> urls = getLinks(html_raw);
+      // choose a link with the highest score
+      PluginPackageLink target;
+      int score = -2;
+      for (auto& item : urls) {
+            if (item.score > score) {
+                  target = item;
+                  score = item.score;
+            }
+      }
+      if (score == -2)
+            return false;
+      if (target.source == GITHUB || target.source == GITHUB_RELEASE) {
+            int release_id;
+            bool has_release = false;
+            if (target.source == GITHUB)
+                  has_release = getLatestRelease(target.user, target.repo, release_id, direct_link);
+            else
+                  has_release = getLatestRelease(target.user, target.repo, release_id, direct_link, QFileInfo(target.url).completeBaseName());
+            if (has_release) {
+                  // there is a GitHub release, change source to RELEASE
+                  if (desc.source != GITHUB_RELEASE || (desc.release_id != release_id && release_id > 0)) {
+                        qDebug() << "desc source original:" << int(desc.source) << " now:release";
+                        qDebug() << "release_id original:" << desc.release_id << "now: " << release_id;
+                        desc.release_id = release_id;
+                        desc.direct_link = direct_link;
+                        desc.source = GITHUB_RELEASE;
+                        return true;
+                  }
+                  else
+                        // already up-to-date
+                        return false;
+            }
+            QString branch;
+            if (target.source == GITHUB_RELEASE)
+                  // for archive links, the URL suffix is the branch name
+                  branch = QFileInfo(target.url).completeBaseName();
+            else
+                  // fetch latest commit sha on master(usually for 3.x), as info for update
+                  branch = target.branch.isNull() ? "master" : target.branch;
```

Comment:
here and above I'd add braces, even if not needed, but there is more than one line